app/src/apdu_handler.c

```diff
@@ -31,12 +31,51 @@
 #include "zxmacros.h"
 
 __Z_INLINE void handleGetPubkey(volatile uint32_t *flags, volatile uint32_t *tx, uint32_t rx) {
-    extractHDPath(rx, OFFSET_DATA);
+    if (rx != 6) {
+        THROW(APDU_CODE_DATA_INVALID);
+    }
 
     uint8_t requireConfirmation = G_io_apdu_buffer[OFFSET_P1];
+    const uint8_t slotIdx = G_io_apdu_buffer[OFFSET_DATA];
+
+    char buffer[20];
+    snprintf(buffer, sizeof(buffer), "%d", slotIdx);
+    zemu_log_stack(buffer);
+
+    account_slot_t slot;
+    zxerr_t err = slot_getSlot(slotIdx,(uint8_t *) &slot, sizeof(slot));
+    snprintf(buffer, sizeof(buffer), "err: %d", err);
+    zemu_log_stack(buffer);
+
+    if (err == zxerr_no_data) {
+        zemu_log_stack("Empty slot");
+        THROW(APDU_CODE_DATA_INVALID);
+    }
+
+    if (err != zxerr_ok) {
+        THROW(APDU_CODE_EXECUTION_ERROR);
+    }
+
+    //We want to reuse extractHDPath, as it works on G_io_apdu_buffer we have to follow 
+    MEMCPY(G_io_apdu_buffer, &slot.path, sizeof(slot.path));
+    rx = sizeof(slot.path);
+    extractHDPath(rx, 0); //this puts hdPath into hdPath global variable
+    *tx = app_fill_address(); //Computes pubkey, the new pubkey is in G_io_apdu_buffer
+
+    //We shift G_io_apdu_buffer  to make space for account
```

Comment:
we control the protocol, shouldn't we first return the address and then account to make this simpler? Alternatively, cannot we change app_fill_address to start from some existing position?

Reply:
Oh. I was not sure if pubkey length is constant. Now I see it is, thus it makes sense to change the order.